### PR TITLE
HPCC-14854 Add Roxie implementation of parallel PROJECT and PARSE

### PR DESCRIPTION
**Pull request by richardkchapman**

### CODE REVIEW

**Comment by richardkchapman** [comment]
@jakesmith Please review.

Note that the full Roxie regression suite has passed with "forceNumStrands=4" which flushed out most of the issues I hope. I have reviewed Gavin's bits and he has reviewed mine, but a third pair of eyes would not hurt.

File: roxie/ccd/ccdactivities.cpp [hunk 1]
```diff
@@ -5164,10 +5165,12 @@ class CRoxieRemoteActivity : public CRoxieSlaveActivity
         {
             remoteGraph->beforeExecute();
             Owned<IFinalRoxieInput> input = remoteGraph->startOutput(0, remoteExtractBuilder.size(), remoteExtractBuilder.getbytes(), false);
-            IEngineRowStream &stream = input->queryStream();
+            Owned <IStrandJunction> junction;
```
Comment (jakesmith) [inline roxie/ccd/ccdactivities.cpp]: trivial: space after Owned

**Comment by jakesmith** [comment]
ccdserver.cpp too big for github to show diffs, so will list line #'s and comment here

trivial : ccdserver (1541)
> // Typically set from hints, common to many stranded activities

Not to all?

**Comment by jakesmith** [comment]
trivial : ccdserver.cpp (1559)
> // Could consider some similar option for numStrands...

Appears to be left over from previous commit, before numStrands option was added

**Comment by jakesmith** [comment]
question : ccdserver.cpp (1749) - CRoxieServerStrandedLateStartActivity

, doesn't look like it's used. What's it intended for?

**Comment by jakesmith** [comment]
question: ccdserver.cpp (1359)
> if (!connected)

Is it okay to connectDependencies() the once? Could the input / streams change, e.g. if conditionals involved ?

**Comment by jakesmith** [comment]
trivial: ccdserver.cpp(6202)

extra blank line

**Comment by jakesmith** [comment]
minor: ccdserver.cpp (19977)
>        : StrandProcessor(_parent, _inputStream, true), helper(_helper)

needsAllocator=true ...

I thought it would need an allocator, but don't see where ParseProcessor uses it.

**Comment by jakesmith** [comment]
style : ccdserver.cpp (19977)
>        : StrandProcessor(_parent, _inputStream, true), helper(_helper)

indented elsewhere.

**Comment by jakesmith** [comment]
@richardkchapman - some minor comments, but looks good.

Now need to start slotting this into Thor.
I wonder how much effort it would be to get base activities similar enough (between Thor and Roxie), so that e.g. StrandProcessor could be common to both engines.

**Comment by richardkchapman** [comment]
A lot, I suspect. Getting the input interface common was hard enough!

I'm not sure more commoning would make sense at the moment. Your ThorDataLink is a little different anyway

**Comment by richardkchapman** [comment]
StrandedLateStart is not used yet but will be used for a stranded filter, for example.

**Comment by ghalliday** [comment]
> Is it okay to connectDependencies() the once? Could the input / streams change, e.g. if conditionals involved ?
Yes - conditionals retain the links to the true and false streams and select between them.

> needsAllocator=true ...
> I thought it would need an allocator, but don't see where ParseProcessor uses it.
Well spotted.  It creates one in the helper classes - and doesn't use this one.  I think it should be addressed in a different PR (I'll open a JIRA HPCC-14876), because it isn't a trivial issue.

**Comment by jakesmith** [comment]
So this PR is ready to merge I think.
@richardkchapman

**Pull request merged**